diff --color -Naru a/includes/deferred/CdnCacheUpdate.php b/includes/deferred/CdnCacheUpdate.php
--- a/includes/deferred/CdnCacheUpdate.php	2023-07-01 00:31:41.203027000 +0800
+++ b/includes/deferred/CdnCacheUpdate.php	2023-07-10 22:00:00.619728633 +0800
@@ -295,14 +295,14 @@
 				'method' => 'PURGE',
 				'url' => $url,
 				'headers' => [
-					'Host' => $urlHost,
+					'Host' => $urlInfo['host'],
 					'Connection' => 'Keep-Alive',
 					'Proxy-Connection' => 'Keep-Alive',
 					'User-Agent' => 'MediaWiki/' . MW_VERSION . ' ' . __CLASS__
 				]
 			];
 			foreach ( $cdnServers as $server ) {
-				$reqs[] = ( $baseReq + [ 'proxy' => $server ] );
+				$reqs[] = ( $baseReq );
 			}
 		}
 
